gk/fib.c

```diff
@@ -0,0 +1,1337 @@
+			}
+		}
+
+		break;
+	}
+
+	case GK_FWD_GATEWAY_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_GATEWAY_BACK_NET: {
+		struct ipaddr *gw_addr =
+			&ip_prefix_fib->u.gateway.ip_addr;
+		struct gk_fib *neigh_fib;
+		struct neighbor_hash_table *neigh_ht;
+		struct ether_cache *eth_cache =
+			ip_prefix_fib->u.gateway.eth_cache;
+
+		eth_cache->ref_cnt--;
+
+		/*
+		 * Find the FIB entry for the gateway.
+		 * We need to release the @eth_cache
+		 * Ethernet header entry from the neighbor hash table.
+		 */
+		if (gw_addr->proto == ETHER_TYPE_IPv4 && gw_addr->ip.v4.s_addr
+				== ip_prefix_addr.ip.v4.s_addr) {
+			/*
+			 * The Ethernet header cache entry is still
+			 * referenced by another prefix, so we cannot
+			 * release this FIB entry.
+			 */
+			if (eth_cache->ref_cnt != 0) {
+				RTE_LOG(ERR, GATEKEEPER,
+					"gk: delete a gateway fib entry [ip prefix = %s], however, its cached Ethernet header is referenced by other entries\n",
+					ip_prefix);
+				return -1;
+			}
+
+			if (ip_prefix_fib->action == GK_FWD_GATEWAY_FRONT_NET)
+				neigh_fib = gk_conf->neigh_fib_front;
+			else
+				neigh_fib = gk_conf->neigh_fib_back;
+
+			neigh_ht = &neigh_fib->u.neigh;
+			ret = rte_hash_del_key(neigh_ht->hash_table,
+				&gw_addr->ip.v4.s_addr);
+			if (ret < 0) {
+				RTE_LOG(ERR, GATEKEEPER,
+					"gk: failed to delete the Ethernet cached header for the IP prefix %s\n",
+					ip_prefix);
+				return -1;
+			}
+
+			/*
+			 * TODO Add concurrency control to
+			 * reset the Ethernet cache entry.
+			 */
+			eth_cache->stale = false;
+			memset(&eth_cache->eth_hdr, 0,
+				sizeof(eth_cache->eth_hdr));
+			eth_cache->ref_cnt = 0;
+		} else if (gw_addr->proto == ETHER_TYPE_IPv6 &&
+				memcmp(gw_addr->ip.v6.s6_addr,
+				ip_prefix_addr.ip.v6.s6_addr,
+				sizeof(gw_addr->ip.v6) == 0)) {
+			/*
+			 * The Ethernet header cache entry is still
+			 * referenced by another prefix, so we cannot
+			 * release this FIB entry.
+			 */
+			if (eth_cache->ref_cnt != 0) {
+				RTE_LOG(ERR, GATEKEEPER,
+					"gk: delete a gateway fib entry [ip prefix = %s], however, its cached Ethernet header is referenced by other entries\n",
+					ip_prefix);
+				return -1;
+			}
+
+			if (ip_prefix_fib->action == GK_FWD_GATEWAY_FRONT_NET)
+				neigh_fib = gk_conf->neigh6_fib_front;
+			else
+				neigh_fib = gk_conf->neigh6_fib_back;
+
+			neigh_ht = &neigh_fib->u.neigh6;
+			ret = rte_hash_del_key(neigh_ht->hash_table,
+				gw_addr->ip.v6.s6_addr);
+			if (ret < 0) {
+				RTE_LOG(ERR, GATEKEEPER,
+					"gk: failed to delete the Ethernet cached header for the IP prefix %s\n",
+					ip_prefix);
+				return -1;
+			}
+
+			/*
+			 * TODO Add concurrency control to
+			 * reset the Ethernet cache entry.
+			 */
+			eth_cache->stale = false;
+			memset(&eth_cache->eth_hdr, 0,
+				sizeof(eth_cache->eth_hdr));
+			eth_cache->ref_cnt = 0;
+		}
+
+		break;
+	}
+
+	case GK_DROP:
+		break;
+
+	case GK_FWD_NEIGHBOR_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_NEIGHBOR_BACK_NET:
```

Comment:
For ``GK_FWD_NEIGHBOR_FRONT_NET`` or ``GK_FWD_NEIGHBOR_BACK_NET`` FIB entries, they are initialized when the Gatekeeper starts. These FIB entries are only reserved for the network prefixes, for which the Gatekeeper is responsible. If we change the network prefixes, then the Gatekeeper may need to restart. That's why I will ignore the deletion of these FIB entries.